File: javaSorting/HeapSort.java
package javaSorting;

public class HeapSort {

	private static void downHeap(int[] data, int size, int index) {
		if (index * 2 + 1 >= size)
			return;
		int left = index * 2 + 1;
		int right = index * 2 + 2;
		if (right == size)
			right -= 1;

		if (data[index] > Math.max(data[left], data[right]))
			return;

		int down_next = left;

		if (data[left] >= data[right] && data[index] < data[left]) {
			int temp = data[left];
			data[left] = data[index];
			data[index] = temp;
		} else if (data[right] > data[left] && data[index] < data[right]) {
			int temp = data[right];
			data[right] = data[index];
			data[index] = temp;
			down_next = right;
		}
		downHeap(data, size, down_next);
	}

	public static void sort(int[] data) {
		int size = data.length;
		for (int i = size / 2 - 1; i >= 0; i--) {
			downHeap(data, size, i);
		}
		for (int i = size - 1; i >= 0; i--) {
			int temp = data[0];
			data[0] = data[i];
			data[i] = temp;
			downHeap(data, i, 0);
		}
	}

	private static void printSortedArray(int data[]) {
		for (int i = 0; i < data.length; i++) {
			System.out.println(data[i]);
		}
	}

	public static void main(String[] args) {
		int SIZE = 1000;
		int[] data = new int[SIZE];
		for (int i = 0; i < SIZE; i++) {
			data[i] = i;
		}

		Shuffle.shuffleData(data);
		sort(data);

		printSortedArray(data);

	}
}

File: Python/heap_sort.py
import time
import random


def _downheap(lista, size, index):
    if index * 2 + 1 >= size: return
    left = index * 2 + 1
    right = index * 2 + 2
    if right == size: right -= 1

    if lista[index] > max(lista[left], lista[right]): return

    down_next = left

    if lista[left] >= lista[right] and lista[index] < lista[left]:
        lista[index], lista[left] = lista[left], lista[index]
    elif lista[right] > lista[left] and lista[index] < lista[right]:
        lista[index], lista[right] = lista[right], lista[index]
        down_next = right

    _downheap(lista, size, down_next)


def heap_sort(lista):
    size = len(lista)
    for i in range(len(lista) // 2 - 1, -1, -1):
        _downheap(lista, size, i)

    for i in range(size - 1, -1, -1): 
        lista[0], lista[i] = lista[i], lista[0]
        _downheap(lista, i, 0)


if __name__ == '__main__':
    lista = [_ for _ in range(1000)]
    random.shuffle(lista)
    start_time = time.time()
    heap_sort(lista)
    end_time = time.time()
    print("Time: ", end_time - start_time)
    print(lista)



